ArduPlane/fence.cpp

```diff
@@ -0,0 +1,132 @@
+            int8_t saved_auto_trim = g.auto_trim;
+            g.auto_trim.set(0);
+            if (fence_act == AC_FENCE_ACTION_RTL_AND_LAND) {
+                set_mode(mode_rtl, ModeReason::FENCE_BREACHED);
+            } else {
+                set_mode(mode_guided, ModeReason::FENCE_BREACHED);
+            }
+            g.auto_trim.set(saved_auto_trim);
+
+            if (fence.get_return_rally() != 0 || fence_act == AC_FENCE_ACTION_RTL_AND_LAND) {
+                guided_WP_loc = rally.calc_best_rally_or_home_location(current_loc, get_RTL_altitude());
+            } else {
+                //return to fence return point, not a rally point
+                guided_WP_loc = {};
+                if (fence.get_return_altitude() > 0) {
+                    // fly to the return point using _retalt
+                    guided_WP_loc.alt = home.alt + 100.0f * fence.get_return_altitude();
+                } else if (fence.get_safe_alt_min() >= fence.get_safe_alt_max()) {
+                    // invalid min/max, use RTL_altitude
+                    guided_WP_loc.alt = home.alt + g.RTL_altitude_cm;
+                } else {
+                    // fly to the return point, with an altitude half way between
+                    // min and max
+                    guided_WP_loc.alt = home.alt + 100.0f * (fence.get_safe_alt_min() + fence.get_safe_alt_max()) / 2;
+                }
+
+                AC_PolyFenceItem return_point;
+                //! TODO: Determine if this works?
+                if (fence.polyfence().get_item(fence.polyfence().num_stored_items() - 1, return_point) &&
+                    return_point.type == AC_PolyFenceType::RETURN_POINT) {
+                    // Fence return point found
+                    guided_WP_loc.lat = return_point.loc.x;
+                    guided_WP_loc.lng = return_point.loc.y;
+                } else {
+                    // Unable to find fence return point
+                    guided_WP_loc.lat = home.lat;
+                    guided_WP_loc.lng = home.lng;
+                }
+            }
+
+
+            //! TODO: Update setting of guided location
+            /*
+                fence_state->guided_lat = guided_WP_loc.lat;
+                fence_state->guided_lng = guided_WP_loc.lng;
+            */
+
+            if (fence.get_action() != AC_FENCE_ACTION_RTL_AND_LAND) {
+                setup_terrain_target_alt(guided_WP_loc);
+                set_guided_WP();
+            }
+
+            if (fence.get_action() == AC_FENCE_ACTION_GUIDED_THROTTLE_PASS) {
+                guided_throttle_passthru = true;
+            }
+            break;
+        }
+
+        AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_FENCE, LogErrorCode(new_breaches));
+    }
+    else if (orig_breaches)
+    {
+        // record clearing of breach
+        AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_FENCE, LogErrorCode::ERROR_RESOLVED);
+    }
+}
+
+bool Plane::fence_stickmixing(void)
```

Comment:
Don't really like how throttle nudging is exempted here.  Would be a lot cleaner if it wasn't - but also a behavioural change.